File: k2/python/host/k2host/fsa_equivalent.py
# Copyright      2020  Xiaomi Corporation (author: Haowen Qiu)

# See ../../../LICENSE for clarification regarding multiple authors
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import torch
from torch.utils.dlpack import to_dlpack

from .fsa import Fsa
from .array import IntArray1
from .array import FloatArray1
from _k2host import IntArray2Size
from _k2host import _RandPath
from _k2host import _is_rand_equivalent
from _k2host import _is_rand_equivalent_max_weight
from _k2host import _is_rand_equivalent_logsum_weight
from _k2host import _is_rand_equivalent_after_rmeps_pruned_logsum


class RandPath(_RandPath):

    def __init__(self, fsa_in: Fsa, no_eps_arc: bool, eps_arc_tries: int = 50):
        super().__init__(fsa_in.get_base(), no_eps_arc, eps_arc_tries)

    def get_sizes(self, array_size: IntArray2Size) -> None:
        return super().get_sizes(array_size)

    def get_output(self, fsa_out: Fsa, arc_map: IntArray1 = None) -> bool:
        return super().get_output(
            fsa_out.get_base(),
            arc_map.get_base() if arc_map is not None else None)


def is_rand_equivalent(fsa_a: Fsa,
                       fsa_b: Fsa,
                       treat_epsilons_specially: bool = True,
                       npath: int = 100) -> bool:
    return _is_rand_equivalent(fsa_a.get_base(), fsa_b.get_base(),
                               treat_epsilons_specially, npath)


def is_rand_equivalent_max_weight(fsa_a: Fsa,
                                  fsa_b: Fsa,
                                  beam: float = float('inf'),
                                  treat_epsilons_specially: bool = True,
                                  delta: float = 1e-6,
                                  top_sorted: bool = True,
                                  npath: int = 100) -> bool:
    return _is_rand_equivalent_max_weight(fsa_a.get_base(), fsa_b.get_base(),
                                          beam, treat_epsilons_specially,
                                          delta, top_sorted, npath)


def is_rand_equivalent_logsum_weight(fsa_a: Fsa,
                                     fsa_b: Fsa,
                                     beam: float = float('inf'),
                                     treat_epsilons_specially: bool = True,
                                     delta: float = 1e-6,
                                     top_sorted: bool = True,
                                     npath: int = 100) -> bool:
    return _is_rand_equivalent_logsum_weight(fsa_a.get_base(),
                                             fsa_b.get_base(), beam,
                                             treat_epsilons_specially, delta,
                                             top_sorted, npath)


def is_rand_equivalent_after_rmeps_pruned_logsum(fsa_a: Fsa,
                                                 fsa_b: Fsa,
                                                 beam: float,
                                                 top_sorted: bool = True,
                                                 npath: int = 100) -> bool:
    return _is_rand_equivalent_after_rmeps_pruned_logsum(
        fsa_a.get_base(), fsa_b.get_base(), beam, top_sorted, npath)

File: k2/python/host/csrc/fsa_equivalent.cc
// k2/python/host/csrc/fsa_equivalent.cc

// Copyright (c)  2020  Xiaomi Corporation (author: Haowen Qiu)

// See ../../../LICENSE for clarification regarding multiple authors

#include "k2/python/host/csrc/fsa_equivalent.h"

#include "k2/csrc/host/array.h"
#include "k2/csrc/host/fsa_equivalent.h"

template <k2host::FbWeightType Type>
void PyBindIsRandEquivalentTpl(py::module &m, const char *name) {
  m.def(
      name,
      [](const k2host::Fsa &a, const k2host::Fsa &b,
         float beam = k2host::kFloatInfinity,
         bool treat_epsilons_specially = true, float delta = 1e-6,
         bool top_sorted = true, std::size_t npath = 100) -> bool {
        return k2host::IsRandEquivalent<Type>(
            a, b, beam, treat_epsilons_specially, delta, top_sorted, npath);
      },
      py::arg("fsa_a"), py::arg("fsa_b"),
      py::arg("beam") = k2host::kFloatInfinity,
      py::arg("treat_epsilons_specially") = true, py::arg("delta") = 1e-6,
      py::arg("top_sorted") = true, py::arg("npath") = 100);
}

void PyBindRandPath(py::module &m) {
  using PyClass = k2host::RandPath;
  py::class_<PyClass>(m, "_RandPath")
      .def(py::init<const k2host::Fsa &, bool, int32_t>(), py::arg("fsa"),
           py::arg("no_eps_arc"), py::arg("eps_arc_tries") = 50)
      .def("get_sizes", &PyClass::GetSizes, py::arg("fsa_size"))
      .def(
          "get_output",
          [](PyClass &self, k2host::Fsa *fsa_out,
             k2host::Array1<int32_t *> *arc_map = nullptr) -> bool {
            return self.GetOutput(fsa_out,
                                  arc_map == nullptr ? nullptr : arc_map->data);
          },
          py::arg("fsa_out"), py::arg("arc_map").none(true));
}

void PybindFsaEquivalent(py::module &m) {
  m.def("_is_rand_equivalent",
        (bool (*)(const k2host::Fsa &, const k2host::Fsa &,
                  bool treat_epsilons_specially, std::size_t)) &
            k2host::IsRandEquivalent,
        py::arg("fsa_a"), py::arg("fsa_b"),
        py::arg("treat_epsilons_specially") = true, py::arg("npath") = 100);

  PyBindIsRandEquivalentTpl<k2host::kMaxWeight>(
      m, "_is_rand_equivalent_max_weight");
  PyBindIsRandEquivalentTpl<k2host::kLogSumWeight>(
      m, "_is_rand_equivalent_logsum_weight");

  // maybe we don't need this version in Python code.
  m.def(
      "_is_rand_equivalent_after_rmeps_pruned_logsum",
      [](const k2host::Fsa &a, const k2host::Fsa &b, float beam,
         bool top_sorted = true, std::size_t npath = 100) -> bool {
        return k2host::IsRandEquivalentAfterRmEpsPrunedLogSum(
            a, b, beam, top_sorted, npath);
      },
      py::arg("fsa_a"), py::arg("fsa_b"), py::arg("beam"),
      py::arg("top_sorted") = true, py::arg("npath") = 100);

  PyBindRandPath(m);
}
